### PR TITLE
Feature/auth basic bearer token validation and generation

### PR DESCRIPTION
**Pull request by varunkumar139**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenService.kt [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+      scopes,
+      nonce,
+    )
+    val accessTokenPayload = accessTokenService.generatePayload(
+      null,
+      null,
+      prisonerData.profile,
+      clientId,
+      scopes,
+      nonce,
```
Comment (tosin-ogunrinde) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenService.kt:111]: access_token does not have a nonce

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenService.kt [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        refreshTokenPayload[v] = claims[value] as Any
+      }
+    } else {
+      refreshTokenPayload = refreshToken.generatePayload(
+        null,
+        null,
+        prisonerData.profile,
+        clientId,
+        scopes,
+        nonce,
```
Comment (tosin-ogunrinde) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenService.kt:128]: refresh_token does not have a nonce

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenGenerationAndValidation.kt [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      payloadMap: HashMap<String, Any>,
+      headerMap: HashMap<String, Any>,
+      alg: SignatureAlgorithm,
+      secret: String,
+    ): String {
+      try {
+        return Jwts.builder()
+          .addClaims(payloadMap)
+          .setHeader(headerMap)
+          .signWith(alg, secret.toByteArray(Charsets.UTF_8))
```
Comment (tosin-ogunrinde) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenGenerationAndValidation.kt:32]: alg is always SignatureAlgorithm.HS256 consider setting it here.

**Review by tosin-ogunrinde — CHANGES_REQUESTED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/resource/UserApprovedClientController.kt [hunk 4]
```diff
@@ -42,7 +53,10 @@ class UserApprovedClientController(
   fun revokeClientAccess(
     @PathVariable("user-id") userId: String,
     @PathVariable("client-id") clientId: UUID,
+    @RequestHeader(HttpHeaders.AUTHORIZATION) authorization: String,
   ): ResponseEntity<Void> {
+    val authenticationInfo = authentication.authenticate(authorization)
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/resource/UserApprovedClientController.kt:58]: These get validated in TokenAuthentication. TokenAuthentication get claim "sub" and "aud" and find client and userapprovedclient.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            throw ApiException(UNAUTHORIZED, UNAUTHORIZED_CODE)
+          }
+        validateScopeInClaims(scopes, userApprovedClient.scopes)
+        return AuthenticationUserInfo(client.id, client.scopes, userId, userApprovedClient.scopes)
+      } else {
+        logger.error("Validation failed for bearer token")
+        throw ApiException(UNAUTHORIZED, UNAUTHORIZED_CODE)
+      }
+    } else {
+      logger.error("Token is not a valid bearer token {}", credential)
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt:69]: This is logging invalid token format

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+          val claims = jwsClaims.body
+          val expireAt = claims["exp"] as Int
+          TokenGenerationAndValidation.validateExpireTime(expireAt)
+          accessTokenId = claims["jti"] as String
+          // verify access token can be cast as UUID
+          UUID.fromString(accessTokenId)
+          clientId = claims["aud"] as String
+          userId = claims["sub"] as String
+          scopes = claims["scopes"] as Any
+        } catch (e: Exception) {
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt:43]: There is no one specific exception in try.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        if (!client.enabled) {
+          logger.warn("Client {} is not enabled", clientId)
+          throw ApiException(UNAUTHORIZED, UNAUTHORIZED_CODE)
+        }
+        val userApprovedClient = userApprovedClientService
+          .getUserApprovedClientByUserIdAndClientId(userId, UUID.fromString(clientId)).orElseThrow {
+            logger.warn("Client {} is not enabled", clientId)
+            throw ApiException(UNAUTHORIZED, UNAUTHORIZED_CODE)
+          }
+        validateScopeInClaims(scopes, userApprovedClient.scopes)
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt:62]: I think we should validate scope. Token is using scope in user approved client for access_token and refresh_token so good to have validate scope. It is standard practise to validate all required claims.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        return AuthenticationUserInfo(client.id, client.scopes, userId, userApprovedClient.scopes)
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt:63]: The standard practise is getting user info from token during authentication and pass to service class for authorisation check if required.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenService.kt [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+import kotlin.collections.LinkedHashMap
+
+const val TOKEN_TYPE = "Bearer"
+
+@Component
+class TokenService(
+  private var prisonerApiService: PrisonerApiService,
+  private var idTokenService: IdTokenPayload,
+  private var accessTokenService: AccessTokenPayload,
+  private var refreshToken: RefreshTokenPayload,
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenService.kt:35]: These classes do not have any private variables so variables will be not shared in threads just like controller or service class as singleton. I agree it is better to initiate new Object each time

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/Authentication.kt [hunk 10]
```diff
@@ -0,0 +1,5 @@
+package uk.gov.justice.digital.hmpps.hmppslaunchpadauth.service.authentication
+
+interface Authentication {
+  fun authenticate(credential: String): AuthenticationInfo
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/Authentication.kt:4]: I think it is better to get client in service class, service class has been written to retrieve data from db. We just need to build auth information object and pass it to service class. Authentication is checked in controller class.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt]: This has been removed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        val accessTokenId: String
+        val scopes: Any
+        try {
+          val jwsClaims = TokenGenerationAndValidation.parseClaims(token, secret)
+          val claims = jwsClaims.body
+          val expireAt = claims["exp"] as Int
+          TokenGenerationAndValidation.validateExpireTime(expireAt)
+          accessTokenId = claims["jti"] as String
+          // verify access token can be cast as UUID
+          UUID.fromString(accessTokenId)
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt]: The current checks for access token and make sure refresh token cannot be used as auth bearer token.
It checks following:
1. jti is valid UUDI
2. ati is not present in claims
3. Expire time is greater than current time
4. sub and aud is present and aud is a valid uuid and sub is valid user format
5. Scopes is present and it contains only scopes from scope list enum set string values

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt]: This has been updated to just check scope list is a valid from scope enum values.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt]: As discussed we only pass client id  after successful login

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenPayload.kt [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    headerClaims["type"] = type
+    return headerClaims
+  }
+
+  protected fun buildCommonClaims(
+    aud: String,
+    sub: String,
+    payloadClaims: LinkedHashMap<String, Any>,
+  ): LinkedHashMap<String, Any> {
+    payloadClaims["iat"] = LocalDateTime.now(ZoneOffset.UTC).toEpochSecond(ZoneOffset.UTC)
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenPayload.kt]: Updated

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/model/AuthorizationGrantType.kt [hunk 13]
```diff
@@ -17,5 +19,14 @@ enum class AuthorizationGrantType(private val grantType: String) {
...
       return false
     }
+
+    fun getAuthorizationGrantTypeByStringValue(value: String): AuthorizationGrantType {
+      AuthorizationGrantType.values().forEach { grantType ->
+        if (value == grantType.toString()) {
+          return grantType
+        }
+      }
+      throw IllegalArgumentException(String.format("Invalid grant value %s", value))
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/model/AuthorizationGrantType.kt:21]: We are using ApiException in service class only method called for serving a request and we know the response code to be sent. For others we should avoid ApiException. These exception will get caught in service class.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenService.kt [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+        logger.warn("Client with id {} not found", clientId)
+        throw ApiException(UNAUTHORIZED, UNAUTHORIZED_CODE)
+      }
+    if (code != null && grantType == AuthorizationGrantType.AUTHORIZATION_CODE.toString() && redirectUri != null) {
+      validateGrant(grantType, clientId, client)
+      val ssoRequest = ssoRequestService.getSsoRequestByAuthorizationCode(code).orElseThrow {
+        logger.warn("Sso Request with code {} not found", code)
+        throw ApiException(UNAUTHORIZED, UNAUTHORIZED_CODE)
+      }
+      validateRedirectUri(redirectUri, ssoRequest)
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenService.kt]: updated

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenService.kt [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+        val clientId = UUID.fromString(clientIdInString)
+        validateGrant(grantType, clientId, client)
+        val userApprovedClient =
+          userApprovedClientService.getUserApprovedClientByUserIdAndClientId(userId, clientId).orElseThrow {
+            logger.warn("User approved client  with user id {} and client id {} not found", userId, clientId)
+            throw ApiException(UNAUTHORIZED, UNAUTHORIZED_CODE)
+          }
+        // add here logic for token not expired
+        if (exp > LocalDateTime.now().toEpochSecond(ZoneOffset.UTC)) {
+          return generateToken(userId, clientId, userApprovedClient.scopes, nonce, claims)
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenService.kt]: This line has been updated. Both access and refresh token contains the scopes which is in user approved clients or sso request.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenService.kt [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+      ssoRequestService.deleteSsoRequestById(ssoRequest.id)
+      return token
+    } else if (refreshToken != null && grantType == "refresh_token") {
+      if (TokenGenerationAndValidation.validateJwtTokenSignature(refreshToken, secret)) {
+        val claims = TokenGenerationAndValidation.parseClaims(refreshToken, secret)
+        val exp = claims.body["exp"] as Int
+        val userId = claims.body["sub"] as String
+        val clientIdInString = claims.body["aud"] as String
+        val clientId = UUID.fromString(clientIdInString)
+        validateGrant(grantType, clientId, client)
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenService.kt]: Yes, these has been added, it make sure jti is valid UUID, ati is not null

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenService.kt [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+        val userApprovedClient =
+          userApprovedClientService.getUserApprovedClientByUserIdAndClientId(userId, clientId).orElseThrow {
+            logger.warn("User approved client  with user id {} and client id {} not found", userId, clientId)
+            throw ApiException(UNAUTHORIZED, UNAUTHORIZED_CODE)
+          }
+        // add here logic for token not expired
+        if (exp > LocalDateTime.now().toEpochSecond(ZoneOffset.UTC)) {
+          return generateToken(userId, clientId, userApprovedClient.scopes, nonce, claims)
+        } else {
+          return generateToken(userId, clientId, userApprovedClient.scopes, nonce, null)
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenService.kt]: Yes, the refresh token when expired will need to have response code Http 400  with error ="expire_refresh_token".
This line has been updated.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenPayload.kt [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    payloadClaims["exp"] = LocalDateTime.now(ZoneOffset.UTC).plusHours(12).toEpochSecond(ZoneOffset.UTC)
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/TokenPayload.kt]: If refresh token in query is valid it just take the payloads and pass payload to generate refresh token class which use existing payload to create refresh token and later new accesstoken id ati is updated only. So refresh token received by using /token and refresh token has everything same except ati. If refresh token has expired then app will need to initaite sso again and get code to get tokens.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/authentication/TokenAuthentication.kt]: Updated with jwt exception

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/IdTokenPayload.kt [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+
+  override fun generatePayload(
+    booking: Booking?,
+    establishment: Establishment?,
+    profile: Profile,
+    clientId: UUID,
+    scopes: Set<Scope>,
+    nonce: String?,
+  ): LinkedHashMap<String, Any> {
+    return buildClaims(booking!!, establishment!!, profile!!, clientId, scopes, nonce)
```
[the remark this answers is not in the document]
Comment (varunkumar139) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppslaunchpadauth/service/token/IdTokenPayload.kt]: this is for id token payload

**Review by tosin-ogunrinde — APPROVED** [review]

**Pull request merged**